### PR TITLE
[FVM] Changed key index type to uint32

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/onflow/flow-go/issues/6204

Most of the changes are only `uint32`->`uint64`, but there are also some other changes:
- The cadence runtime interface is still using `uint64` the converion happens in the FVM. This will get removed once cadence is also updated.
- The storage used migration also migrates all account statuses to newer versions (by reading them, and writing them back)

> [!NOTE]  
> The emulator still needs to be updated for the CI to pass

### CODE REVIEW

File: fvm/environment/accounts_status.go [hunk 1]
```diff
@@ -70,27 +82,90 @@ func (a *AccountStatus) ToBytes() []byte {
...
 	}
 
-	if len(inp) != accountStatusSize {
+	// this is to migrate old account status to new account status on the fly
+	// TODO: remove this whole block after Crescendo, when a full migration will be made.
+	if len(inp) == accountStatusSizeV2 {
+		// migrate v2 to v3
+
+		inp2 := make([]byte, accountStatusSizeV2)
+		copy(inp2, inp)
```
Comment (peterargue) [inline fvm/environment/accounts_status.go:110]: is this `copy()` needed? I think the append on line 134 is copying the data a second time.

File: fvm/environment/accounts_status_test.go [hunk 2]
```diff
@@ -54,13 +54,36 @@ func TestAccountStatus(t *testing.T) {
...
+		oldBytes := []byte{
+			0,                      // flags
+			0, 0, 0, 0, 0, 0, 0, 7, // storage used
+			0, 0, 0, 0, 0, 0, 0, 6, // storage index
+			0, 0, 0, 0, 0, 0, 0, 5, // public key counts
+			0, 0, 0, 0, 0, 0, 0, 3, // account id counter
+		}
+
+		// The new format has an extra 8 bytes for the account id counter
+		// so we need to increase the storage used by 8 bytes while migrating it
```
Comment (peterargue) [inline fvm/environment/accounts_status_test.go:79]:
```suggestion
		// for v2->v3 migration, we are shrinking the public key counts from uint64 to uint32
		// so we need to decrease the storage used by 4 bytes
```

File: cmd/util/ledger/migrations/storage_used_migration.go [hunk 3]
```diff
@@ -82,58 +90,57 @@ func (m *AccountUsageMigration) MigrateAccount(
...
+	newStatusValue := status.ToBytes()
+	statusSizeDiff := len(newStatusValue) - len(statusValue)
+
+	// the status size diff should be added to the actual size
+	if statusSizeDiff < 0 {
+		if uint64(-statusSizeDiff) > actualUsed {
+			log.Error().
+				Str("account", address.HexWithPrefix()).
+				Msgf("account storage used would be negative")
+			return nil
```
Comment (peterargue) [inline cmd/util/ledger/migrations/storage_used_migration.go:107]: should this return an error?

**Review by peterargue — COMMENTED** [review]
thanks for cleaning this all up.

File: fvm/environment/accounts_status.go (hunk 1 above)
Comment (janezpodhostnik) [reply fvm/environment/accounts_status.go:110]: yes it. Slicing `inp` in place caused issues. I added a comment

**Review by ramtinms — APPROVED** [review]
The changes looks good to me, though since this is changing both the models.Flow definition of accounts and some of the access endpoint, Its a big breaking change, we need to make sure all the dependencies (e.g. flow portal, ledger, ... ) will still work. 
My suggestion is to only enforce this limit inside FVM and eventually move towards the full change.

**Comment by janezpodhostnik** [comment]
@ramtinms the grpc endpoints already have the key index as uint32, so most sdks would not need to change at all. The flow-go sdk just needed to change, because we import the types from there into flow-go. 

Do you think there could still be impacts on dependencies?

**Comment by ramtinms** [comment]
The last time we changed the model files, there were some 3rd party dependencies on it (I think it was Coinbase thing). If the GRPCs are already uint32 then we just need to give heads up if anyone is importing the model file directly.

**Comment by turbolent** [comment]
I now have opened #6220 to update to the latest version of Cadence, which includes onflow/cadence#3465.

Do we want to merge these to PRs together?

**Comment by turbolent** [comment]
Looks like we need https://github.com/onflow/flow-go-sdk/pull/705 first

**Comment by turbolent** [comment]
@janezpodhostnik I merged everything together in https://github.com/onflow/flow-go/compare/bastian/change-key-index-to-uint32:
- This PR
- Update to latest SDK which has the key index changes
- Update to latest Cadence which has the key index changes

**Review by peterargue — APPROVED** [review]

**Pull request merged**